sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CatalogFileIndex.scala

```diff
@@ -67,7 +67,10 @@ class CatalogFileIndex(
       val selectedPartitions = sparkSession.sessionState.catalog.listPartitionsByFilter(
         table.identifier, filters)
       val partitions = selectedPartitions.map { p =>
-        PartitionPath(p.toRow(partitionSchema), p.storage.locationUri.get)
+        val path = new Path(p.storage.locationUri.get)
+        val fs = path.getFileSystem(hadoopConf)
+        PartitionPath(
+          p.toRow(partitionSchema), path.makeQualified(fs.getUri, fs.getWorkingDirectory))
```

Comment:
But we will qualify it before writing to it at [here](https://github.com/apache/spark/pull/15705/files#diff-f9858c2d9d1a3c3e48753ef675bc865aR69), doesn't it work?